pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -227,11 +240,22 @@ func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitR
 				Admit:   false,
 			}
 		}
-		hints[container.Name] = result
-	}
 
-	m.podTopologyHints[string(pod.UID)] = hints
-	klog.Infof("[topologymanager] Topology Affinity for Pod: %v are %v", pod.UID, m.podTopologyHints[string(pod.UID)])
+		klog.Infof("[topologymanager] Topology Affinity for (pod: %v container: %v): %v", pod.UID, container.Name, result)
+		if m.podTopologyHints[string(pod.UID)] == nil {
+			m.podTopologyHints[string(pod.UID)] = make(map[string]TopologyHint)
+		}
+		m.podTopologyHints[string(pod.UID)][container.Name] = result
+
+		err := m.allocateAlignedResources(pod, &container)
```

Comment:
It seems that topology manager callout resource managers to pre-allocate compute resources here.
But I cannot find a logic to free pre-allocated resources when one of containers got error from `m.calculateAffinity()`. in this PR.
I think we should free pre-reserved resources in that case.
(Maybe it is handled by some logic... like `RemoveContainer`)

So, I`m curious that what will happen on the below situation.
-4 containers in a Pod.
-The first contaienr is admitted and resources are allocated(by `allocateAlignedResources`).
-The Second container got error from `m.calculateAffinity` and admission is rejected.

Reply:
It is not really correct to say "the first container is admitted". 

The split of the original `Allocate` into two parts: `Allocate` and `AddContainer` means that in reality the "pre-allocation" is only acted upon when the pod is admitted. 

In your scenario, the pod will not be admitted due to the second container's `m.calculateAffinity` error.
Therefore the resources which were "pre-allocated" to the first container will not be consumed and those resources remain free.

Reply:
`It is not really correct to say "the first container is admitted".`
`The split of the original AddContainer into two parts`
=> I understood these, points.

`in reality the "pre-allocation" is only acted upon when the pod is admitted.`
=> But I didnt understand this point, because it seems CPU Manager(static policy) allocates cpus for container on `Allocate` part right after hint calculation of a container, so I thought cpus will be allocated even Pod is not admitted.

When the returned value of admit in below code is ture, Topology Manager calls out `Allocate` funtion of resource managers by `allocateAlignedResources()`.
`result, admit := m.calculateAffinity(pod, &container)`

Here, in the case of CPU Manager, CPUManager calls out `staticPolicy.Allocate(..)`.
(I looked only CPU Manager.)

Then It seems static policy : 
1) allocates cpus based on stored affinity for container by `staticPolicy.allocateCPUs() and takeByTopology() internally`
2) update `state` of cpus to exclude allocated cpus by the below code.
    `s.SetDefaultCPUSet(s.GetDefaultCPUSet().Difference(result))`

Reply:
It's possible that CPUs may get allocated even if the pod is ultimately not admitted. However, any cpus allocated to non-admitted pods (as well as any pods that may have terminated for some other reason unbeknownst to the kubelet) are lazily reclaimed at the top of `GetTopologyHints()` here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L303

Reply:
Oh, I missed that garbage collector, I will take a look it.
Thankyou @nolancon @klueska

Reply:
I've tested out your scenario locally @bg-chun and it works as expected in @klueska's description.